File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-eq-shared-5-4/Correlation.cpp
// Correlation.cpp 
#include "Correlation.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Correlation::Correlation(TiXmlNode* xmlNode)
: CalculationFromObservation(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //notionalAmountNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* notionalAmountNode = xmlNode->FirstChildElement("notionalAmount");

   if(notionalAmountNode){notionalAmountIsNull_ = false;}
   else{notionalAmountIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- notionalAmountNode , address : " << notionalAmountNode << std::endl;
   #endif
   if(notionalAmountNode)
   {
      if(notionalAmountNode->Attribute("href") || notionalAmountNode->Attribute("id"))
      {
          if(notionalAmountNode->Attribute("id"))
          {
             notionalAmountIDRef_ = notionalAmountNode->Attribute("id");
             notionalAmount_ = boost::shared_ptr<NonNegativeMoney>(new NonNegativeMoney(notionalAmountNode));
             notionalAmount_->setID(notionalAmountIDRef_);
             IDManager::instance().setID(notionalAmountIDRef_,notionalAmount_);
          }
          else if(notionalAmountNode->Attribute("href")) { notionalAmountIDRef_ = notionalAmountNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { notionalAmount_ = boost::shared_ptr<NonNegativeMoney>(new NonNegativeMoney(notionalAmountNode));}
   }

   //correlationStrikePriceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* correlationStrikePriceNode = xmlNode->FirstChildElement("correlationStrikePrice");

   if(correlationStrikePriceNode){correlationStrikePriceIsNull_ = false;}
   else{correlationStrikePriceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- correlationStrikePriceNode , address : " << correlationStrikePriceNode << std::endl;
   #endif
   if(correlationStrikePriceNode)
   {
      if(correlationStrikePriceNode->Attribute("href") || correlationStrikePriceNode->Attribute("id"))
      {
          if(correlationStrikePriceNode->Attribute("id"))
          {
             correlationStrikePriceIDRef_ = correlationStrikePriceNode->Attribute("id");
             correlationStrikePrice_ = boost::shared_ptr<CorrelationValue>(new CorrelationValue(correlationStrikePriceNode));
             correlationStrikePrice_->setID(correlationStrikePriceIDRef_);
             IDManager::instance().setID(correlationStrikePriceIDRef_,correlationStrikePrice_);
          }
          else if(correlationStrikePriceNode->Attribute("href")) { correlationStrikePriceIDRef_ = correlationStrikePriceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { correlationStrikePrice_ = boost::shared_ptr<CorrelationValue>(new CorrelationValue(correlationStrikePriceNode));}
   }

   //boundedCorrelationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* boundedCorrelationNode = xmlNode->FirstChildElement("boundedCorrelation");

   if(boundedCorrelationNode){boundedCorrelationIsNull_ = false;}
   else{boundedCorrelationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- boundedCorrelationNode , address : " << boundedCorrelationNode << std::endl;
   #endif
   if(boundedCorrelationNode)
   {
      if(boundedCorrelationNode->Attribute("href") || boundedCorrelationNode->Attribute("id"))
      {
          if(boundedCorrelationNode->Attribute("id"))
          {
             boundedCorrelationIDRef_ = boundedCorrelationNode->Attribute("id");
             boundedCorrelation_ = boost::shared_ptr<BoundedCorrelation>(new BoundedCorrelation(boundedCorrelationNode));
             boundedCorrelation_->setID(boundedCorrelationIDRef_);
             IDManager::instance().setID(boundedCorrelationIDRef_,boundedCorrelation_);
          }
          else if(boundedCorrelationNode->Attribute("href")) { boundedCorrelationIDRef_ = boundedCorrelationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { boundedCorrelation_ = boost::shared_ptr<BoundedCorrelation>(new BoundedCorrelation(boundedCorrelationNode));}
   }

   //numberOfDataSeriesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* numberOfDataSeriesNode = xmlNode->FirstChildElement("numberOfDataSeries");

   if(numberOfDataSeriesNode){numberOfDataSeriesIsNull_ = false;}
   else{numberOfDataSeriesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- numberOfDataSeriesNode , address : " << numberOfDataSeriesNode << std::endl;
   #endif
   if(numberOfDataSeriesNode)
   {
      if(numberOfDataSeriesNode->Attribute("href") || numberOfDataSeriesNode->Attribute("id"))
      {
          if(numberOfDataSeriesNode->Attribute("id"))
          {
             numberOfDataSeriesIDRef_ = numberOfDataSeriesNode->Attribute("id");
             numberOfDataSeries_ = boost::shared_ptr<XsdTypePositiveInteger>(new XsdTypePositiveInteger(numberOfDataSeriesNode));
             numberOfDataSeries_->setID(numberOfDataSeriesIDRef_);
             IDManager::instance().setID(numberOfDataSeriesIDRef_,numberOfDataSeries_);
          }
          else if(numberOfDataSeriesNode->Attribute("href")) { numberOfDataSeriesIDRef_ = numberOfDataSeriesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { numberOfDataSeries_ = boost::shared_ptr<XsdTypePositiveInteger>(new XsdTypePositiveInteger(numberOfDataSeriesNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<NonNegativeMoney> Correlation::getNotionalAmount()
{
   if(!this->notionalAmountIsNull_){
        if(notionalAmountIDRef_ != ""){
             return boost::shared_static_cast<NonNegativeMoney>(IDManager::instance().getID(notionalAmountIDRef_));
        }else{
             return this->notionalAmount_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<NonNegativeMoney>();
   }
}
boost::shared_ptr<CorrelationValue> Correlation::getCorrelationStrikePrice()
{
   if(!this->correlationStrikePriceIsNull_){
        if(correlationStrikePriceIDRef_ != ""){
             return boost::shared_static_cast<CorrelationValue>(IDManager::instance().getID(correlationStrikePriceIDRef_));
        }else{
             return this->correlationStrikePrice_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CorrelationValue>();
   }
}
boost::shared_ptr<BoundedCorrelation> Correlation::getBoundedCorrelation()
{
   if(!this->boundedCorrelationIsNull_){
        if(boundedCorrelationIDRef_ != ""){
             return boost::shared_static_cast<BoundedCorrelation>(IDManager::instance().getID(boundedCorrelationIDRef_));
        }else{
             return this->boundedCorrelation_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BoundedCorrelation>();
   }
}
boost::shared_ptr<XsdTypePositiveInteger> Correlation::getNumberOfDataSeries()
{
   if(!this->numberOfDataSeriesIsNull_){
        if(numberOfDataSeriesIDRef_ != ""){
             return boost::shared_static_cast<XsdTypePositiveInteger>(IDManager::instance().getID(numberOfDataSeriesIDRef_));
        }else{
             return this->numberOfDataSeries_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypePositiveInteger>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-eq-shared-5-4/Correlation.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Correlation : CalculationFromObservation
    {
        public Correlation(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode notionalAmountNode = xmlNode.SelectSingleNode("notionalAmount");
            
            if (notionalAmountNode != null)
            {
                if (notionalAmountNode.Attributes["href"] != null || notionalAmountNode.Attributes["id"] != null) 
                {
                    if (notionalAmountNode.Attributes["id"] != null) 
                    {
                        notionalAmountIDRef_ = notionalAmountNode.Attributes["id"].Value;
                        NonNegativeMoney ob = new NonNegativeMoney(notionalAmountNode);
                        IDManager.SetID(notionalAmountIDRef_, ob);
                    }
                    else if (notionalAmountNode.Attributes["href"] != null)
                    {
                        notionalAmountIDRef_ = notionalAmountNode.Attributes["href"].Value;
                    }
                    else
                    {
                        notionalAmount_ = new NonNegativeMoney(notionalAmountNode);
                    }
                }
                else
                {
                    notionalAmount_ = new NonNegativeMoney(notionalAmountNode);
                }
            }
            
        
            XmlNode correlationStrikePriceNode = xmlNode.SelectSingleNode("correlationStrikePrice");
            
            if (correlationStrikePriceNode != null)
            {
                if (correlationStrikePriceNode.Attributes["href"] != null || correlationStrikePriceNode.Attributes["id"] != null) 
                {
                    if (correlationStrikePriceNode.Attributes["id"] != null) 
                    {
                        correlationStrikePriceIDRef_ = correlationStrikePriceNode.Attributes["id"].Value;
                        CorrelationValue ob = new CorrelationValue(correlationStrikePriceNode);
                        IDManager.SetID(correlationStrikePriceIDRef_, ob);
                    }
                    else if (correlationStrikePriceNode.Attributes["href"] != null)
                    {
                        correlationStrikePriceIDRef_ = correlationStrikePriceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        correlationStrikePrice_ = new CorrelationValue(correlationStrikePriceNode);
                    }
                }
                else
                {
                    correlationStrikePrice_ = new CorrelationValue(correlationStrikePriceNode);
                }
            }
            
        
            XmlNode boundedCorrelationNode = xmlNode.SelectSingleNode("boundedCorrelation");
            
            if (boundedCorrelationNode != null)
            {
                if (boundedCorrelationNode.Attributes["href"] != null || boundedCorrelationNode.Attributes["id"] != null) 
                {
                    if (boundedCorrelationNode.Attributes["id"] != null) 
                    {
                        boundedCorrelationIDRef_ = boundedCorrelationNode.Attributes["id"].Value;
                        BoundedCorrelation ob = new BoundedCorrelation(boundedCorrelationNode);
                        IDManager.SetID(boundedCorrelationIDRef_, ob);
                    }
                    else if (boundedCorrelationNode.Attributes["href"] != null)
                    {
                        boundedCorrelationIDRef_ = boundedCorrelationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        boundedCorrelation_ = new BoundedCorrelation(boundedCorrelationNode);
                    }
                }
                else
                {
                    boundedCorrelation_ = new BoundedCorrelation(boundedCorrelationNode);
                }
            }
            
        
            XmlNode numberOfDataSeriesNode = xmlNode.SelectSingleNode("numberOfDataSeries");
            
            if (numberOfDataSeriesNode != null)
            {
                if (numberOfDataSeriesNode.Attributes["href"] != null || numberOfDataSeriesNode.Attributes["id"] != null) 
                {
                    if (numberOfDataSeriesNode.Attributes["id"] != null) 
                    {
                        numberOfDataSeriesIDRef_ = numberOfDataSeriesNode.Attributes["id"].Value;
                        XsdTypePositiveInteger ob = new XsdTypePositiveInteger(numberOfDataSeriesNode);
                        IDManager.SetID(numberOfDataSeriesIDRef_, ob);
                    }
                    else if (numberOfDataSeriesNode.Attributes["href"] != null)
                    {
                        numberOfDataSeriesIDRef_ = numberOfDataSeriesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        numberOfDataSeries_ = new XsdTypePositiveInteger(numberOfDataSeriesNode);
                    }
                }
                else
                {
                    numberOfDataSeries_ = new XsdTypePositiveInteger(numberOfDataSeriesNode);
                }
            }
            
        
        }
        
    
        #region NotionalAmount_
        private NonNegativeMoney notionalAmount_;
        public NonNegativeMoney NotionalAmount_
        {
            get
            {
                if (this.notionalAmount_ != null)
                {
                    return this.notionalAmount_; 
                }
                else if (this.notionalAmountIDRef_ != null)
                {
                    notionalAmount_ = IDManager.getID(notionalAmountIDRef_) as NonNegativeMoney;
                    return this.notionalAmount_; 
                }
                else
                {
                      return this.notionalAmount_; 
                }
            }
            set
            {
                if (this.notionalAmount_ != value)
                {
                    this.notionalAmount_ = value;
                }
            }
        }
        #endregion
        
        public string notionalAmountIDRef_ { get; set; }
        #region CorrelationStrikePrice_
        private CorrelationValue correlationStrikePrice_;
        public CorrelationValue CorrelationStrikePrice_
        {
            get
            {
                if (this.correlationStrikePrice_ != null)
                {
                    return this.correlationStrikePrice_; 
                }
                else if (this.correlationStrikePriceIDRef_ != null)
                {
                    correlationStrikePrice_ = IDManager.getID(correlationStrikePriceIDRef_) as CorrelationValue;
                    return this.correlationStrikePrice_; 
                }
                else
                {
                      return this.correlationStrikePrice_; 
                }
            }
            set
            {
                if (this.correlationStrikePrice_ != value)
                {
                    this.correlationStrikePrice_ = value;
                }
            }
        }
        #endregion
        
        public string correlationStrikePriceIDRef_ { get; set; }
        #region BoundedCorrelation_
        private BoundedCorrelation boundedCorrelation_;
        public BoundedCorrelation BoundedCorrelation_
        {
            get
            {
                if (this.boundedCorrelation_ != null)
                {
                    return this.boundedCorrelation_; 
                }
                else if (this.boundedCorrelationIDRef_ != null)
                {
                    boundedCorrelation_ = IDManager.getID(boundedCorrelationIDRef_) as BoundedCorrelation;
                    return this.boundedCorrelation_; 
                }
                else
                {
                      return this.boundedCorrelation_; 
                }
            }
            set
            {
                if (this.boundedCorrelation_ != value)
                {
                    this.boundedCorrelation_ = value;
                }
            }
        }
        #endregion
        
        public string boundedCorrelationIDRef_ { get; set; }
        #region NumberOfDataSeries_
        private XsdTypePositiveInteger numberOfDataSeries_;
        public XsdTypePositiveInteger NumberOfDataSeries_
        {
            get
            {
                if (this.numberOfDataSeries_ != null)
                {
                    return this.numberOfDataSeries_; 
                }
                else if (this.numberOfDataSeriesIDRef_ != null)
                {
                    numberOfDataSeries_ = IDManager.getID(numberOfDataSeriesIDRef_) as XsdTypePositiveInteger;
                    return this.numberOfDataSeries_; 
                }
                else
                {
                      return this.numberOfDataSeries_; 
                }
            }
            set
            {
                if (this.numberOfDataSeries_ != value)
                {
                    this.numberOfDataSeries_ = value;
                }
            }
        }
        #endregion
        
        public string numberOfDataSeriesIDRef_ { get; set; }
        
    
        
    
    }
    
}

